chroma-common/blockdevices/blockdevice_zfs.py

```diff
@@ -78,14 +88,50 @@ def __exit__(self, type, value, traceback):
             self.unlock_pool()
 
     def lock_pool(self):
-        lock = self.import_locks[self.pool_path]
-        lock.acquire()
-        # log.debug('ZfsDevice acquired lock on %s' % self.pool_path)
+        """
+        Attempt to acquire a lock on a given zpool through a lockfile, increment the reference count each time
+        this method is called, regardless of whether acquire is actually called. Keep trying until we know
+        we are locking the zpool.
+
+        Lock acquire polls the timeout at intervals of LOCK_ACQUIRE_TIMEOUT.
+        """
+        if ZfsDevice.locks_dir_initialized == False:
+            # ensure lock directory exists
+            try:
+                os.mkdir(self.ZPOOL_LOCK_DIR)
+            except OSError:
+                pass
+
+            ZfsDevice.locks_dir_initialized = True
+
+        while not self.lock.i_am_locking():
+            try:
+                self.lock.acquire(self.LOCK_ACQUIRE_TIMEOUT)
+                with open(self.lock.lock_file, 'w') as f:
+                    f.writelines(str(self.lock.pid))
+            except LockTimeout:
+                # prune locks owned by non-existent processes
+                with open(self.lock.lock_file, 'r') as f:
+                    contents = f.readlines()
+
+                assert len(contents) == 1 and contents[0].isdigit(), \
+                    'unexpected contents of lockfile %s: %s' % (self.lock.lock_file, contents)
+
+                # validate pid holding lock exists
+                if not pid_exists(int(contents[0])):
+                    self.lock.break_lock()
+
+        self.lock_refcount[self.lock_unique_id] += 1
 
     def unlock_pool(self):
-        lock = self.import_locks[self.pool_path]
-        lock.release()
-        # log.debug('ZfsDevice released lock on %s' % self.pool_path)
+        """
+        Release the held lock, should only be called when holding the lock. Decrement the reference count
+        each time this method is called and release when reference count is equal to one.
+        """
+        self.lock_refcount[self.lock_unique_id] -= 1
```

Comment:
yes it is, I would suggest moving it out into its own file in a separate patch (as this patch should be concerned with replacing the existing locking mechanism , in-place)